subsys/dfu/fmfu_fdev/src/fmfu_fdev.c

```diff
@@ -0,0 +1,255 @@
+		uint32_t read_len = MIN(NRF_FMFU_MODEM_BUFFER_SIZE, bytes_left);
+
+		err = flash_read(fdev, read_addr, buf, read_len);
+		if (err != 0) {
+			LOG_ERR("flash_read failed2: %d", err);
+			return err;
+		}
+
+		err = nrf_fmfu_memory_chunk_write(seg_target_addr, read_len,
+						  buf);
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_memory_chunk_write failed: %d", err);
+			return err;
+		}
+
+		LOG_DBG("wrote chunk: offset 0x%x target addr 0x%x size 0x%x",
+			read_addr, seg_target_addr, read_len);
+
+		seg_target_addr += read_len;
+		bytes_left -= read_len;
+		read_addr += read_len;
+	}
+
+	return 0;
+}
+
+static int prevalidate(uint8_t *buf, size_t wrapper_len)
+{
+	int err;
+
+	err = nrf_fmfu_verify_signature(wrapper_len, (void *)buf);
+	if (err != 0) {
+		LOG_ERR("nrf_fmfu_verify_signature failed: %d", errno);
+	}
+
+	return err;
+}
+
+static int load_segments(const struct device *fdev, uint8_t *meta_buf,
+			 size_t wrapper_len, const Segments_t *seg,
+			 size_t blob_offset, uint8_t *buf)
+{
+	int err;
+	size_t prev_segments_len = 0;
+
+	LOG_INF("Writing %d segments", seg->_Segments__Segment_count);
+
+	for (int i = 0; i < seg->_Segments__Segment_count; i++) {
+		size_t seg_size = seg->_Segments__Segment[i]._Segment_len;
+		uint32_t seg_addr =
+			seg->_Segments__Segment[i]._Segment_target_addr;
+		uint32_t read_addr = blob_offset + prev_segments_len;
+
+		err = nrf_fmfu_transfer_start();
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_transfer_start failed: %d", err);
+			return err;
+		}
+
+		err = load_segment(fdev, seg_size, seg_addr, read_addr, buf);
+		if (err != 0) {
+			LOG_ERR("load_segment failed: %d", err);
+			return err;
+		}
+
+		LOG_DBG("Modem state: %d", nrf_fmfu_modem_state_get());
+
+		err = nrf_fmfu_transfer_end();
+		if (err != 0) {
+			LOG_ERR("nrf_fmfu_transfer_end failed, state: %d errno: %d",
+				nrf_fmfu_modem_state_get(), errno);
+			return err;
+		}
+
+		if (i == 0) {
+			/* This is the first segment, which contains the
+			 * IPC-DFU bootloader. We can now perform the
+			 * signature verification.
+			 */
+			err = prevalidate(meta_buf, wrapper_len);
+			if (err != 0) {
+				LOG_ERR("Prevalidate failed: %d", err);
+				return err;
+			}
+		}
+
+		prev_segments_len += seg_size;
+
+		LOG_INF("Segment %d written. Target addr: 0x%x, size: 0%x", i,
+			seg_addr, seg_size);
+	}
+
+	LOG_INF("FMFU finished");
+
+	return 0;
+}
+
+int fmfu_fdev_load(uint8_t *buf, size_t buf_len, const struct device *fdev,
+		   size_t offset)
+{
+	const cbor_string_type_t *segments_string;
+	COSE_Sign1_Manifest_t wrapper;
+	bool hash_len_valid = false;
+	uint8_t expected_hash[32];
+	bool hash_valid = false;
+	Segments_t segments;
+	size_t blob_offset;
+	size_t wrapper_len;
+	uint8_t hash[32];
+	size_t blob_len;
+	uint32_t unused;
+	int err;
+
+	if (buf == NULL || fdev == NULL) {
+		return -ENOMEM;
+	}
+
+	/* Put modem in DFU/RPC state */
+	err = nrf_fmfu_init(NULL, &unused);
+	if (err != 0) {
+		LOG_ERR("nrf_fmfu_init failed, errno: %d\n.", errno);
```

Comment:
`err` instead of `errno`

Reply:
`errno` is a weird concept when it comes to bsdlib which is globally available, and its set to whatever return value the lib gives us, so this is actually correct. 

This is (AFAIK) the idiomatic way of reporting errors from this API.

Reply:
Oohh I was wondering if this was the same mechanism as Linux's `<errno.h>`, but I didn't think Zephyr used that. Didn't realize bsdlib also utilized this.

My comments about this can be ignored then.